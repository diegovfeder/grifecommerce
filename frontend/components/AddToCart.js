import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import { CURRENT_USER_QUERY } from './User';

const ADD_TO_CART_MUTATION = gql`
	mutation ADD_TO_CART_MUTATION($id: ID!) {
		addToCart(productId: $id) {
			id
		}
	}
`;

// eslint-disable-next-line react/prop-types
export default function AddToCart({ id }) {
	const [addToCart, { loading }] = useMutation(ADD_TO_CART_MUTATION, {
		variables: { id },
		refetchQueries: [{ query: CURRENT_USER_QUERY }],
	});
	return (
		<button disabled={loading} type="button" onClick={addToCart}>
			Add{loading && 'ing'} To Cart ðŸ›’
		</button>
	);
}
