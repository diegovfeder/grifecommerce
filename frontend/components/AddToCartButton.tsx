
import React, {ReactNode} from 'react';
import styled from 'styled-components';
import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import { CURRENT_USER_QUERY } from '../hooks/useUserQuery';

interface AddToCartButtonProps {
  id: String;
	children?: ReactNode;
}

const StyledButtonLink = styled.button`
	color: #393939;
	&:hover {
		cursor: pointer;
		text-decoration: underline;
	}
`;

// TODO: Properly type this
const update = (cache: any, payload: any) => {
	cache.evict(cache.identify(payload.data.deleteProduct));
};

const ADD_TO_CART_MUTATION = gql`
	mutation ADD_TO_CART_MUTATION($id: ID!) {
		addToCart(productId: $id) {
			id
		}
	}
`;

const AddToCartButton = ({ id, children }: AddToCartButtonProps) => {
	const [addToCart, { loading }] = useMutation(ADD_TO_CART_MUTATION, {
		variables: { id },
		refetchQueries: [{ query: CURRENT_USER_QUERY }],
	});

	return (
		<StyledButtonLink
			type="button"
			disabled={loading}
			onClick={() => addToCart()}
		>
			{children ? children : undefined}
		</StyledButtonLink>
	);
}

export default AddToCartButton;
