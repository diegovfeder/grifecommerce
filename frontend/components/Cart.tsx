import styled from 'styled-components';
import StyledCart from './styles/StyledCart';
import CloseButton from './styles/CloseButton';
import Supreme from './styles/Supreme';
import formatMoney from '../util/formatMoney';
import { useUser } from './User';
import calcTotalPrice from '../util/calcTotalPrice';
import { useCart } from '../util/cartState';
import RemoveFromCart from './RemoveFromCart';
import { Checkout } from './Checkout';
import { CartItemProps } from '../types/commonTypes';

const CartItemStyles = styled.li`
	padding: 1rem 0;
	border-bottom: 1px solid var(--lightGrey);
	display: grid;
	grid-template-columns: auto 1fr auto;
	img {
		margin-right: 1rem;
	}
	h3,
	p {
		margin: 0;
	}
`;

const CartItem = ({ id, quantity, product }: CartItemProps) => {
	if (!product) return null;
	return (
		<CartItemStyles>
			<img
				width="100"
				src={product.photo.image.publicUrlTransformed}
				alt={product.name}
			/>
			<div>
				<h3>{product.name}</h3>
				<p>
					{formatMoney(product.price * quantity)}-
					<em>
						{quantity} &times; {formatMoney(product.price)} each
					</em>
				</p>
			</div>
			<RemoveFromCart id={id} />
		</CartItemStyles>
	);
};

const Cart = () => {
	const me = useUser();
	const { cartOpen, closeCart } = useCart();
	if (!me) return null;
	return (
		<StyledCart open={cartOpen}>
			<header>
				<Supreme>{me.name}'s Cart</Supreme>
				<CloseButton onClick={closeCart}>&times;</CloseButton>
			</header>
			<ul>
				{me.cart.map((cartItem: CartItemProps) => (
					<CartItem key={cartItem.id} {...cartItem} />
				))}
			</ul>
			<footer>
				<p>{formatMoney(calcTotalPrice(me.cart))}</p>
				<Checkout />
			</footer>
		</StyledCart>
	);
};

export default Cart;
