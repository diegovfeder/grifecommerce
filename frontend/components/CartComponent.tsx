import useCartState from '../hooks/useCartState';
import useUserQuery from '../hooks/useUserQuery';
import formatMoney from '../utils/formatMoney';
import calcTotalPrice from '../utils/calcTotalPrice';
import { CartItemProps } from '../types/commonTypes';
import Supreme from './styles/Supreme';
import StyledCart from './styles/StyledCart';
import StyledCloseButton from './styles/StyledCloseButton';
import StyledCartItem from './styles/StyledCartItem';
import Checkout from './Checkout';
import RemoveFromCartButton from './RemoveFromCartButton';

const CartItem = ({ id, quantity, product }: CartItemProps) => {
	if (!product) return null;
	return (
		<StyledCartItem>
			<img
				width="100"
				src={product?.photo?.image?.publicUrlTransformed}
				alt={product?.name}
			/>
			<div>
				<h3>{product.name}</h3>
				<p>
					{formatMoney(product.price * quantity)} -{' '}
					<em>
						{quantity} &times; {formatMoney(product.price)} each
					</em>
				</p>
			</div>
			<RemoveFromCartButton id={id} />
		</StyledCartItem>
	);
};

const CartComponent = () => {
	const user = useUserQuery();
	const { cartOpen, closeCart } = useCartState();
	if (!user) return null;
	return (
		<StyledCart open={cartOpen}>
			<header>
				<Supreme>{user.name}'s Cart</Supreme>
				<StyledCloseButton onClick={closeCart}>&times;</StyledCloseButton>
			</header>
			<ul>
				{user.cart.map((cartItem: CartItemProps) => (
					<CartItem key={cartItem.id} {...cartItem} />
				))}
			</ul>
			<footer>
				<p>{formatMoney(calcTotalPrice(user.cart))}</p>
				<Checkout />
			</footer>
		</StyledCart>
	);
};

export default CartComponent;
