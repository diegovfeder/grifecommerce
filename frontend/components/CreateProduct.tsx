import { useMutation } from '@apollo/client';
import Router from 'next/router';
import { useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import { object, string, number, mixed } from 'yup';

import { CREATE_PRODUCT_MUTATION } from '../gql/mutations';
import ALL_PRODUCTS_QUERY from '../gql/queryAllProducts.gql';
import StyledForm from './styles/StyledForm';
import ErrorMessage from './error/ErrorMessage';

type FormInputs = {
	name: string;
	description: string;
	price: number;
	photo: string;
};

const schema = object({
	name: string().required(),
	description: string().optional(),
	price: number().positive().integer().required(),
	photo: mixed().required('You need to provide a file').default(undefined),
}).required();

const CreateProduct = () => {
	const {
		handleSubmit,
		register,
		reset,
		formState: { errors },
	} = useForm<FormInputs>({
		resolver: yupResolver(schema),
	});

	const [createProductImage] = useMutation(CREATE_PRODUCT_MUTATION, {
		variables: {
			image: '',
			altText: 'test',
		},
		onCompleted: data => {
			console.log({ data });
		},
		onError: err => {
			console.error({ err });
		},
	});

	const [createProduct, { loading, error }] = useMutation(
		CREATE_PRODUCT_MUTATION,
		{
			variables: {
				name: 'Cool Shoes',
				description: 'These are the coolest shoes',
				price: 5000,
				photo: {
					image: 'cool-shoes.jpg',
					altText: 'Cool Shoes',
				},
				status: 'AVAILABLE',
			},
			refetchQueries: [{ query: ALL_PRODUCTS_QUERY }],
			onCompleted: data => {
				console.log({ data });
				reset();
				// Router.push({
				// 	pathname: `/product/${data?.createProduct?.id || ''}`,
				// });
			},
			onError: err => {
				console.error({ err });
			},
		},
	);

	// FIXME:
	const onSubmit = async (data: FormInputs) => {
		console.log(data);
		// const productImage = await createProductImage({
		// 	variables: {
		// 		image: data.photo,
		// 		altText: data.name,
		// 		// product: probably create product first? but pass in the id
		// 	},
		// });
		// console.log({ productImage });

		const product = await createProduct({
			variables: {
				name: data.name,
				description: data.description,
				price: data.price,
				photo: {
					image: data.photo[0],
					altText: data.name,
				},
				status: 'AVAILABLE',
			},
		});
		console.log({ product });
	};

	return (
		<StyledForm onSubmit={handleSubmit(onSubmit)}>
			<ErrorMessage error={error} />
			{errors.photo && <p>{errors.photo.message}</p>}
			<fieldset disabled={loading} aria-busy={loading}>
				<label
					style={{ display: 'flex', flex: 1, flexDirection: 'column' }}
					htmlFor="name"
				>
					Name
					<input
						{...register('name')}
						id="name"
						type="text"
						placeholder="Name"
					/>
				</label>
				<label
					style={{ display: 'flex', flex: 1, flexDirection: 'column' }}
					htmlFor="description"
				>
					Description
					<textarea
						{...register('description')}
						id="description"
						placeholder="Description"
					/>
				</label>
				<label
					style={{ display: 'flex', flex: 1, flexDirection: 'column' }}
					htmlFor="price"
				>
					Price
					<input
						{...register('price')}
						id="price"
						type="number"
						placeholder="price"
					/>
				</label>
				<label
					style={{ display: 'flex', flex: 1, flexDirection: 'column' }}
					htmlFor="photo"
				>
					Photo
					<input
						{...register('photo')}
						id="photo"
						type="file"
						placeholder="photo"
						accept=".jpg, .png, .gif, .jpeg"
					/>
				</label>

				<button type="submit">+ Add Product</button>
			</fieldset>
		</StyledForm>
	);
};

export default CreateProduct;
