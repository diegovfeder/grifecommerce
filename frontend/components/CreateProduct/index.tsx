import { useMutation } from '@apollo/client';
import { gql } from 'graphql-tag';
import useForm from 'hooks/useForm';
import StyledForm from 'components/styles/StyledForm';
import ErrorMessage from 'components/ErrorMessage';
import { Container } from './styles';

const CREATE_PRODUCT_MUTATION = gql`
	mutation CREATE_PRODUCT_MUTATION(
		$name: String!
		$description: String!
		$price: Int!
		$image: Upload
	) {
		createProduct(
			data: {
				name: $name
				description: $description
				price: $price
				status: "AVAILABLE"
				photo: { create: { image: $image, altText: $name } }
			}
		) {
			id
			name
			description
			price
		}
	}
`;

function CreateProduct() {
	const { inputs, handleChange, clearForm, resetForm } = useForm({
		name: '',
		price: undefined,
		description: '',
	});

	const [createProduct, { loading, error, data }] = useMutation(
		CREATE_PRODUCT_MUTATION,
		{
			variables: inputs,
			// refetchQueries: [{ query: ALL_PRODUCTS_QUERY }],
		},
	);

	return (
		<Container>
			<StyledForm
				onSubmit={async (e: { preventDefault: () => void }) => {
					e.preventDefault();
					console.log(inputs);
					// Submit inputFields to backend
					const res = await createProduct();
					clearForm();
					// Router.push({
					//   pathname: `/product/${res.data.createProduct.id}`,
					// });
				}}
			>
				<ErrorMessage error={error} />
				<fieldset disabled={loading} aria-busy={loading}>
					<label htmlFor="name">
						Image
						<input
							type="file"
							id="image"
							name="image"
							onChange={handleChange}
						/>
					</label>
					<label htmlFor="name">
						Name
						<input
							type="text"
							id="name"
							name="name"
							placeholder="Name"
							value={inputs.name}
							onChange={handleChange}
						/>
					</label>
					<label htmlFor="description">
						Description
						<input
							type="text"
							id="description"
							name="description"
							placeholder="description"
							value={inputs.description}
							onChange={handleChange}
						/>
					</label>
					<label htmlFor="price">
						Price
						<input
							type="number"
							id="price"
							name="price"
							placeholder="Price"
							value={inputs.price}
							onChange={handleChange}
						/>
					</label>
					<button type="submit">+ Add Product</button>
				</fieldset>
			</StyledForm>
		</Container>
	);
}

export default CreateProduct;
