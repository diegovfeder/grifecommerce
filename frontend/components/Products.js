import { useQuery } from '@apollo/client';
import gql from 'graphql-tag';
import styled from 'styled-components';
import { totalProductsPerPage } from '../config';
import Product from './Product';

export const ALL_PRODUCTS_QUERY = gql`
	query ALL_PRODUCTS_QUERY($take: Int, $skip: Int = 0) {
		products(take: $take, skip: $skip) {
			id
			name
			price
			description
			photo {
				id
				image {
					publicUrlTransformed
				}
			}
		}
	}
`;

const ProductsListStyles = styled.div`
	display: grid;
	grid-template-columns: 1fr 1fr;
	grid-gap: 60px;
`;

// eslint-disable-next-line react/prop-types
export default function Products({ page }) {
	const { data, error, loading } = useQuery(ALL_PRODUCTS_QUERY, {
		variables: {
			take: totalProductsPerPage,
			skip: page * totalProductsPerPage - totalProductsPerPage,
		},
	});
	if (loading) return <p>Loading...</p>;
	if (error) return <p>Error: {error.message}</p>;
	return (
		<div>
			<ProductsListStyles>
				{data.products.map((product) => (
					<Product key={product.id} product={product} />
				))}
			</ProductsListStyles>
		</div>
	);
}
