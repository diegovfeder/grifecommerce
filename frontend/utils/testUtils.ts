// eslint-disable-next-line import/no-extraneous-dependencies
import casual from 'casual';
import { PRODUCTS_COUNT_QUERY } from '../gql/queries';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
	// __typename: 'Item',
	id: '718b7ac6-7cf1-47e7-b1de-c4a13ca92f2d',
	name: 'GRIFE Sample Pack - (Yellow)',
	description:
		'This our first sample pack for summer beats in a sunny day, when all is yellow :)',
	price: 5000,
	// user: null,
	photo: {
		// FIXME:
		// id: 'abc123',
		// altText: 'dogs are best',
		image: {
			publicUrlTransformed: '/samplepack_yellow.jpg',
		},
	},
});

const fakeUser = (rest?: any) => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: [],
	...rest,
});

const fakeOrderItem = () => ({
	__typename: 'OrderItem',
	id: casual.uuid,
	image: {
		image: `${casual.word}.jpg`,
	},
	name: casual.words(),
	price: 4234,
	quantity: 1,
	description: casual.words(),
});

const fakeOrder = () => ({
	__typename: 'Order',
	id: 'ord123',
	charge: 'ch_123',
	total: 40000,
	items: [fakeOrderItem(), fakeOrderItem()],
	createdAt: '2022-12-11T20:16:13.797Z',
	user: fakeUser(),
});

const fakeCartItem = (rest?: any) => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 3,
	product: fakeItem(),
	user: fakeUser(),
	...rest,
});

// FIXME: Properly type this class
// Fake LocalStorage
class LocalStorageMock {
	store: { [key: string]: any } = {};
	constructor() {
		this.store = {};
	}

	clear() {
		this.store = {};
	}

	getItem(key: string) {
		return this.store[key] || null;
	}

	setItem(key: string, value: string) {
		this.store[key] = value.toString();
	}

	removeItem(key: string) {
		delete this.store[key];
	}
}

function makePaginationMocksFor(length: number) {
	return [
		{
			request: { query: PRODUCTS_COUNT_QUERY },
			result: {
				data: {
					_allProductsMeta: {
						count: length,
					},
					itemsConnection: {
						__typename: 'aggregate',
						aggregate: {
							count: length,
							__typename: 'count',
						},
					},
				},
			},
		},
	];
}

export {
	makePaginationMocksFor,
	LocalStorageMock,
	fakeItem,
	fakeUser,
	fakeCartItem,
	fakeOrder,
	fakeOrderItem,
};
