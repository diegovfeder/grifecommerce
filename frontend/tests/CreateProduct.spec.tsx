import { render, screen, waitFor } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import userEvent from '@testing-library/user-event';
import Router from 'next/router'; // We will MOCK THIS
import wait from 'waait';
import CreateProduct, {
	CREATE_PRODUCT_MUTATION,
} from '../components/CreateProduct';
<<<<<<< HEAD:frontend/tests/CreateProduct.test.js
import { fakeItem, makePaginationMocksFor } from '../utils/testUtils';
import { PRODUCTS_QUERY } from '../components/Products';
=======
import {
	fakeItem,
	// makePaginationMocksFor
} from '../utils/testUtils';
import PRODUCTS_QUERY from '../gql/queryProducts.gql';
>>>>>>> dev:frontend/tests/CreateProduct.spec.tsx

import { logRoles } from '@testing-library/dom';

// TODO: Enforce typing for this fakeItem();
const item = fakeItem();

jest.mock('next/router', () => ({
	push: jest.fn(),
}));

const mocks = [
	{
		request: {
			query: PRODUCTS_QUERY,
			variables: {
				take: 1,
				skip: 0,
			},
		},
		result: {
			data: {
				products: [item],
				productsCount: 1,
			},
		},
	},
	{
		request: {
			query: PRODUCTS_QUERY,
			variables: {
				take: 4,
				skip: 0,
			},
		},
		result: {
			data: {
				products: [item, item, item, item],
				productsCount: 4,
			},
		},
	},
];

describe('<CreateProduct/>', () => {
	it('renders and matches snapshot', () => {
		const {
			container,
			// debug
		} = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<CreateProduct />
			</MockedProvider>,
		);
		expect(container).toMatchSnapshot();
	});

	it('handles the updating', async () => {
		const { container } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<CreateProduct />
			</MockedProvider>,
		);
		logRoles(container);

		// 2. type into the boxes
		await userEvent.type(screen.getByPlaceholderText(/Name/i), item.name);
		await userEvent.type(
			screen.getByPlaceholderText(/Price/i),
			item.price.toString(),
		);
		await userEvent.type(
			screen.getByPlaceholderText(/Description/i),
			item.description,
		);

		// 3.  check that those boxes are populated!
		expect(screen.getByDisplayValue(item.name)).toBeInTheDocument();
		expect(screen.getByDisplayValue(item.price)).toBeInTheDocument();
		expect(screen.getByDisplayValue(item.description)).toBeInTheDocument();
	});

	it('creates the items when the form is submitted', async () => {
		const mocks = [
			{
				request: {
					query: CREATE_PRODUCT_MUTATION,
					variables: {
						name: item.name,
						description: item.description,
						image: '',
						price: item.price,
					},
				},
				result: {
					data: {
						createProduct: {
							...item, // TODO: Pass down all fake item fields
							id: 'abc123',
						},
					},
				},
			},
			{
				request: {
					query: PRODUCTS_QUERY,
					variables: { take: 2, skip: 0 },
				},
				result: {
					data: {
						allProducts: [item],
					},
				},
			},
		];

		const { container } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<CreateProduct />
			</MockedProvider>,
		);
		logRoles(container);

		// Type into the inputs
		// 2. type into the boxes
		await userEvent.type(screen.getByPlaceholderText(/Name/i), item.name);
		await userEvent.type(
			screen.getByPlaceholderText(/Price/i),
			item.price.toString(),
		);
		await userEvent.type(
			screen.getByPlaceholderText(/Description/i),
			item.description,
		);

		// Submit it and see if the page change has been called
		await userEvent.click(screen.getByText(/Add Product/));
		await waitFor(() => wait(0));
		expect(Router.push).toHaveBeenCalled();

		// FIXME: Check this
		expect(Router.push).toHaveBeenCalledWith({ pathname: '/product/abc123' });
	});
});
