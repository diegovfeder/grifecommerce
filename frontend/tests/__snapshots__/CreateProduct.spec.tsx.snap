// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateProduct/> renders and matches snapshot 1`] = `
<div>
  <form
    class="StyledForm-sc-44wenk-0 gkQMPk"
  >
    <fieldset
      aria-busy="false"
    >
      <label
        for="name"
        style="display: flex; flex: 1; flex-direction: column;"
      >
        Name
        <input
          id="name"
          name="name"
          placeholder="Name"
          type="text"
          value=""
        />
      </label>
      <label
        for="description"
        style="display: flex; flex: 1; flex-direction: column;"
      >
        Description
        <textarea
          id="description"
          name="description"
          placeholder="Description"
        />
      </label>
      <label
        for="price"
        style="display: flex; flex: 1; flex-direction: column;"
      >
        Price
        <input
          id="price"
          name="price"
          placeholder="price"
          type="number"
          value="0"
        />
      </label>
      <label
        for="image"
      >
        Image
        <input
          id="image"
          name="image"
          required=""
          type="file"
        />
      </label>
      <button
        type="submit"
      >
        + Add Product
      </button>
    </fieldset>
  </form>
</div>
`;
